src/subtype.c

```diff
@@ -4304,6 +4304,211 @@ int jl_subtype_matching(jl_value_t *a, jl_value_t *b, jl_svec_t **penv)
+                if (body != newbody)
+                    newbody = jl_new_struct(jl_unionall_type, var, newbody);
+                // n.b. we do not widen lb, since that would be the wrong direction
+                newvar = insert_nondiagonal(var->ub, troot, widen2ub);
+                if (newvar != var->ub) {
+                    newvar = (jl_value_t*)jl_new_typevar(var->name, var->lb, newvar);
+                    newbody = jl_apply_type1(newbody, newvar);
+                    newbody = jl_type_unionall((jl_tvar_t*)newvar, newbody);
+                }
+            }
+            type = newbody;
+            JL_GC_POP();
+        }
+    }
+    else if (jl_is_uniontype(type)) {
+        jl_value_t *a = ((jl_uniontype_t*)type)->a;
+        jl_value_t *b = ((jl_uniontype_t*)type)->b;
+        jl_value_t *newa = NULL;
+        jl_value_t *newb = NULL;
+        JL_GC_PUSH2(&newa, &newb);
+        newa = insert_nondiagonal(a, troot, widen2ub);
+        newb = insert_nondiagonal(b, troot, widen2ub);
+        if (newa != a || newb != b)
+            type = jl_new_struct(jl_uniontype_type, newa, newb);
+        JL_GC_POP();
+    }
+    else if (jl_is_vararg(type)) {
+        // As for Vararg we'd better widen it's var to ub as otherwise they are still diagonal
+        jl_value_t *t = jl_unwrap_vararg(type);
+        jl_value_t *n = jl_unwrap_vararg_num(type);
+        widen2ub = !(n && jl_is_long(n)) || jl_unbox_long(n) > 1;
+        jl_value_t *newt;
+        JL_GC_PUSH1(&newt);
+        newt = insert_nondiagonal(t, troot, widen2ub);
+        if (t != newt)
+            type = jl_new_struct(jl_vararg_type, newt, n);
+        JL_GC_POP();
+    }
+    else if (jl_is_datatype(type)) {
+        if (jl_is_tuple_type(type)) {
+            jl_svec_t *newparams = NULL;
+            JL_GC_PUSH1(&newparams);
+            for (size_t i = 0; i < jl_nparams(type); i++) {
+                jl_value_t *elt = jl_tparam(type, i);
+                jl_value_t *newelt = insert_nondiagonal(elt, troot, widen2ub);
+                if (elt != newelt) {
+                    if (!newparams) {
+                        newparams = (jl_svec_t*)newelt; // temporary root
+                        newparams = jl_svec_copy(((jl_datatype_t*)type)->parameters);
+                    }
+                    jl_svecset(newparams, i, newelt);
+                }
+            }
+            if (newparams)
+                type = (jl_value_t*)jl_apply_tuple_type(newparams, 1);
+            JL_GC_POP();
+        }
+    }
+    return type;
+}
+
+static jl_value_t *_widen_diagonal(jl_value_t *t, jl_varbinding_t *troot) {
+    check_diagonal(t, troot, 0);
+    int any_concrete = 0;
+    for (jl_varbinding_t *v = troot; v != NULL; v = v->prev) {
+        v->concrete = v->occurs_cov > 1 && v->occurs_inv == 0;
+        any_concrete |= v->concrete;
+    }
+    if (!any_concrete)
+        return t; // no diagonal
+    return insert_nondiagonal(t, troot, 0);
```

Comment:
Since you already checked occurs_inv, it may be already fine to do `wident2ub` always?
```suggestion
    return insert_nondiagonal(t, troot, 1);
```
In effect though, that may be the same as implementing it as a map of `rewrap_unionall` over the fields (but recursively) roughly doing `map(rewrap_unionall, t.parameters, Repeated(u))`?

I think the tricky cases I was thinking of were possibly something like:
`Tuple{Val{S}, T, T} where {S, T<:S}` or perhaps
`Tuple{S, T, T} where {S, T<:S}` or perhaps
`Tuple{S, Vararg{T}} where {S, T<:S}`

These could be safely widened to `Tuple{Val, Any, Any}`, `Tuple{Any, Any, Any}`, and `Tuple{Any, Vararg{Any}}` respectively, but I was hoping to instead convert them to the more precise results of `Tuple{Val{S}, T1, T2} where {S, T1<:S, T2<:S}`, `Tuple{S, T1, T2} where {S, T1<:S, T2<:S}`, and `Tuple{S, Vararg{Union{T1,T2} where T1<:S where T2<:S}} where S`

(though normally allocation normalization will have incorrectly broken that last type anyways, since it quickly converts that to `Tuple{S, Vararg{T where T<:S}} where S` which it thinks is equal to `Tuple{S, Vararg{S}} where S` even though it is not necessarily correct to replace T.ub with T in the expression `T where T` there–as that also implies a new constraint on the `Vararg`: that it is a diagonal Vararg when it was not intended to be)

Reply:
It seems ok that
```julia
julia> CC.instanceof_tfunc(Type{Tuple{S, Vararg{T}}} where {S, T<:S})[1]
Tuple{S, Vararg{S}} where S
```
As `S` should be always concrete at runtime and `Tuple{S, Vararg{S}} where S` should cover all possible types.

The current code fails for this case though
```julia
julia> CC.instanceof_tfunc(Type{Tuple{S, S, Vararg{T}}} where {S, T<:S})[1]
Tuple{S2, S1, Vararg{S}} where {S, S1<:S, S2<:S}
```
It shows that the `widen2ub` branch should be recursive, and return `Tuple{S2, S1, Vararg{Any}} where {S, S1<:S, S2<:S}` instead.